vtgate_gateway_collectd.py

```diff
@@ -0,0 +1,58 @@
+#!/usr/bin/python
+
+import time
+import util
+import mock
+
+NAME = 'vtgate'
+
+#Module that generates metric using the vtgate /debug/gateways endpoint
+#Metrics generated
+# For all keyspace shards  servingMaster: 1/0 indicating whether there is a Master tablet up and serving
+class VtgateGateway(util.BaseCollector):
+    def __init__(self, collectd, json_provider=None, verbose=False, interval=None):
+        super(VtgateGateway, self).__init__(collectd, NAME, 15001, json_provider, verbose, interval)
+
+    def configure_callback(self, conf):
+        super(VtgateGateway, self).configure_callback(conf)
+        self.register_read_callback()
+
+    def process_data(self, json_data):
+        keyspaces = group_tablets_by_keyspace(json_data)
+
+        for keyspaceName, tablets in keyspaces.items():
+            foundServingMaster = False
+            for tabletsForType in tablets:
+                if "MASTER" == tabletsForType["tabletType"]:
+                    for tablet in tabletsForType['tabletsStats']:
+                        if tablet['Up'] and tablet['Serving']:
+                            foundServingMaster = True
+                            break
+            self.emitter.emit("servingMaster", 1 if foundServingMaster else 0, 'gauge', {"ks": tabletsForType["keyspace"], "shard": tabletsForType["shard"]})
+
+
+def group_tablets_by_keyspace(json_data):
+    keyspaces = {}
+    for keyspaceShardTabletType, tablets in json_data.items():
```

Comment:
might wanna wrap the inner part of this loop in a `try .... except Exception: ...` and log the error because it seems like there's potentially a ton of reasons why the loop might fail (thinking fields missing from your json and you trying to access a subfield etc) and you don't wanna fail all the other keyspaces / the whole plugin because parsing 1 of them fails.